sale_mrp_bom/models/sale_order_line.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2020 Akretion Renato Lima <renato.lima@akretion.com.br>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    bom_id = fields.Many2one(
+        comodel_name="mrp.bom",
+        string="BoM",
+        domain="[('product_tmpl_id.product_variant_ids', '=', product_id),"
+        "'|', ('product_id', '=', product_id), "
+        "('product_id', '=', False)]",
+    )
+
+    @api.constrains("bom_id", "product_id")
+    def _check_match_product_variant_ids(self):
+        for line in self:
+            if line.bom_id:
+                bom_product_tmpl = line.bom_id.product_tmpl_id
+                bom_product = bom_product_tmpl.product_variant_ids
+            else:
+                bom_product_tmpl, bom_product = None, None
```

Comment:
Constraint is incorrect, can be a lot more simple:
```

    @api.constrains("bom_id", "product_id")
    def _check_match_product_id(self):
        for line in self:
            if line.product_id != line.bom_id.product_id:
                raise ValidationError(
                    _(
                        "Please select BoM that has matched product with the line `{}`"
                    ).format(line.product_id.name)
                )

```

Reply:
Agree that the code looks all wrong, but don't we want to allow BoMs that only specify a product template and no specific variant? It was recently fixed in 14.0 like that: https://github.com/OCA/sale-workflow/pull/2947/files#diff-c6c892ccbd2c51f5c8a648ee366fd8c32f9ecd9ebce7c649b8c48f1f29aad900

Reply:
https://github.com/OCA/sale-workflow/pull/2947/files we will forward port this